haystack/preview/components/builders/dynamic_prompt_builder.py

```diff
@@ -0,0 +1,334 @@
+        else:
+            raise ValueError(
+                f"{self.__class__.__name__} was not provided with a list of ChatMessage(s) or a string template."
+                "Please check the parameters passed to its run method."
+            )
+        return {"prompt": result}
+
+    def _process_simple_template(self, prompt_source: str, template_variables: Dict[str, Any]) -> str:
+        """
+        Renders the template from the provided string source with the provided template variables.
+
+        :param prompt_source: A Jinja2 template as a string.
+        :type prompt_source: str
+        :param template_variables: A dictionary of template variables.
+        :type template_variables: Dict[str, Any]
+        :return: A string containing the rendered template.
+        :rtype: str
+        """
+        template = self._validate_template(prompt_source, set(template_variables.keys()))
+        return template.render(template_variables)
+
+    def _process_chat_messages(self, prompt_source: List[ChatMessage], template_variables: Dict[str, Any]):
+        """
+        Processes a list of :class:`ChatMessage` instances to generate a chat prompt.
+
+        It takes the last user message in the list, treats it as a template, and renders it with the provided
+        template variables. The resulting message replaces the last user message in the list, forming a complete,
+        templated chat prompt.
+
+        :param prompt_source: A list of `ChatMessage` instances to be processed. The last message is expected
+        to be from a user and is treated as a template.
+        :type prompt_source: List[ChatMessage]
+
+        :param template_variables: A dictionary of template variables used for rendering the last user message.
+        :type template_variables: Dict[str, Any]
+
+        :return: A list of `ChatMessage` instances, where the last user message has been replaced with its
+        templated version.
+        :rtype: List[ChatMessage]
+
+        :raises ValueError: If `chat_messages` is empty or contains elements that are not instances of
+        `ChatMessage`.
+        :raises ValueError: If the last message in `chat_messages` is not from a user.
+        """
+        if not prompt_source:
+            raise ValueError(
+                f"The {self.__class__.__name__} requires a non-empty list of ChatMessage instances. "
+                f"Please provide a valid list of ChatMessage instances to render the prompt."
+            )
+        if not all(isinstance(message, ChatMessage) for message in prompt_source):
+            raise ValueError(
+                f"The {self.__class__.__name__} expects a list containing only ChatMessage instances. "
+                f"The provided list contains other types. Please ensure that all elements in the list "
+                f"are ChatMessage instances."
+            )
+
+        last_message: ChatMessage = prompt_source[-1]
+        if last_message.is_from(ChatRole.USER):
```

Comment:
Why do we enforce this? Not sure whether it makes sense to use a template for another role though. Just thinking. 🤔

Reply:
For now, it is simple; later we can adapt or add some callable that applies custom rule.